### ISSUE TITLE
animations: initial push for animations

### ISSUE THREAD

**Issue by pkozlowski-opensource** [ng-bootstrap/ng-bootstrap#295]
Some of the BS components need animations beyond simple CSS3 transitions (those are taken care of by BS classes). Components that need non-trivial animations are:
- [ ] Accordion
- [x] Alert => #1800
- [ ] Carousel
- [ ] Collapse
- [ ] Popover
- [ ] Modal
- [x] Tabset => #1802
- [ ] Tooltip

**Comment by Foxandxss** [comment]
Animations that needs the @host can't be done as today until https://github.com/angular/angular/issues/9044 is fixed.

**Comment by pkozlowski-opensource** [comment]
> Animations that needs the @host can't be done as today

Yeh, run into this one very quickly after playing with animations for a sec... Marking it as blocked on the upstream issues...

**Comment by pkozlowski-opensource** [comment]
Blocked on https://github.com/angular/angular/issues/9044

**Comment by aciccarello** [comment]
FYI, angular/angular#9044 was closed on July 14th

**Comment by pkozlowski-opensource** [comment]
@aciccarello unfortunately  angular/angular#9044 added support for animations in `@Component` host but not on `@Directive`, sadly...

**Comment by aciccarello** [comment]
Ah, so you must be waiting on angular/angular#9947 then.

**Comment by pkozlowski-opensource** [comment]
> Ah, so you must be waiting on angular/angular#9947 then.

Yep.

**Comment by rolandoldengarm** [comment]
Ah I thought they would do that in RC5?

Can you guys add the carousel to the list as well?

**Comment by wesleycho** [comment]
@StanFaas please read the discussion & links provided.

An update will be posted if the status changes - consider silence to mean that this is still blocked.

**Comment by colinmorelli** [comment]
Correct me if I'm wrong here, but I don't believe all of these items are blocked by the angular issue. Modals, for example, are based off of `@Component` which does have rich animation support. We should be able to support at least a stock animation for modal windows based on what already exists, no?

**Comment by pkozlowski-opensource** [comment]
@colinmorelli you are totally right, not all animations / widgets are blocked. In fact I had a chat with Matias recently and IMO we are only blocked on:
* accordion, carousel and tabs that need `query` support
* collapse that need support for animations in directive hosts

Certain basic fade / slide animations should be doable right now so in theory it should possible to cover: Alert, Modal, Popover / Tooltip. It is just that till now we were focusing on stabilizing APIs to minimize breaking changes in the future. 

If anyone is willing to step in and give Alert, Modal, Popover / Tooltip a go it would be more than appreciated!

**Comment by colinmorelli** [comment]
I'm interested in your opinion on how animations should be handled @pkozlowski-opensource 

As far as I'm aware, animations by adding/removing CSS classes is not yet supported in angular (would love to be wrong on this). Assuming that's the case, replicating the stock `fade in` animations that are present on most components with the angular DSL is trivial; but allowing users to customize would be harder.

Any suggestions on how we could be more flexible with animations before we get started on this?

**Comment by pkozlowski-opensource** [comment]
> As far as I'm aware, animating by adding/removing CSS classes is not yet supported in angular (would love to be wrong on this). 

Correct.

> Assuming that's the case, replicating the stock fade in animations that are present on most components with the angular DSL is trivial; 

IMO replicating BS animations is the only reasonable option we've got now if we want to use built-in angular animations.

> but allowing users to customize would be harder.

Yes, but:
* most of those animations don't require anything that would be critical for theming (fonts, colors, spacing etc. - only duration and the effect itself)
* Matias was thinking of some mechanism to share / override animation definitions

> Any suggestions on how we could be more flexible with animations before we get started on this

Honestly at this point animations are so new in Angular and not many people got any experience with it. So the best suggestion for now would be: start exploring and see how far we can go, nothing down any roadblocks along the way. Based on what works and what doesn't we could provide feedback to Matias and drive future changes in the animate module.

**Comment by pkozlowski-opensource** [comment]
OK, so it looks like animations are getting mature enough to start incorporating them into ng-bootstrap :+1:  
Here is a simple example: http://plnkr.co/edit/a3SO8Aqb5g6xXQHev3AW?p=preview

Before we go full steam ahead here are the items that need investigation:
* [ ] impact of animations on code size (weight of a polyfill etc.)
* [ ] impact on tests (ex.: do we need to change our tests and if so - how?)
* [ ] impact on browsers support (what is going to happen in ie9?)
* [ ] impact on server-side rendering and AoT (everything still works?)

**Comment by wesleycho** [comment]
The only polyfill needed for animations is the web animations polyfill.

**Comment by intellix** [comment]
any further idea of how you guys want to implement this? Was interested in getting animations for my modals and looking into how to do it here

**Comment by masaanli** [comment]
Waiting for this one too

**Comment by MickL** [comment]
Do we really need to reimplement all animations? I mean Bootstrap only adds simple classes and the animations are already part of the css/scss. We just have to adapt the same add/remove class behaviour like Bootstrap JS does.

**Comment by colinmorelli** [comment]
@MickL Angular2 does not currently support adding/removing classes with the animations builder (I'm really not sure why here, do you know @pkozlowski-opensource)

Effectively, this means that while animating in would be easy (we could just add a class to the element in the next run loop), animating out would be a pain. The issue is that, without angular supporting an animated class removal, we would not know when the animation is done and the component should be removed from the DOM tree. The only real solution here would be to force a no-op animation with a duration that exactly matches the CSS animation, and remove the class in ngOnDestroy. However that would lead to strange timing issues.

For those reasons, I think bootstrap animations will likely need to be recreated with the angular animations framework. It's unfortunately going to make customization a lot harder.

**Comment by MickL** [comment]
Why use the animations builder when the animations are already done in css by Bootstrap? Just add the classes, no need for Angular animations

**Comment by MickL** [comment]
Bootstrap is adding the classes by JS, too. No need to do it in another or better way.

**Comment by colinmorelli** [comment]
It's not about doing it in another or a better way, it's just about doing it in a way that works with the Angular component lifecycle. Without declaring an animation on a component (using angular's animations framework), destroying a component removes it from the DOM immediately. `ngOnDestroy()` would not work as a place to remove a class and have it be animated.

Using modals as an example, what we could do here - albeit by adding a separate animations process - is have the `NgbModal` service call a method on the modal window component, then bind a listener to the `transitionend` event, at which point the `NgbModal` service could actually destroy the component.

@MickL just to be clear - I'm personally 100% in support of doing this with CSS class manipulation if possible. In addition to requiring less code in the framework, it's also far more customizable. It's just not as straight forward to do in angular because of the component lifecycle issue. Unless I'm missing something.

**Comment by maxisam** [comment]
as for "impact on browsers support (what is going to happen in ie9?)"

I created a small library to detect if the browser is supported. If not, turn off animation, so everything will still work. Under the hood, it uses user agent to detect browsers.

Here is the plunker. 

http://plnkr.co/edit/VQAZFUcWYbdsDnqjjPHX?p=preview

**Comment by ymeine** [comment]
Hi everyone, 

I tried to tackle this issue by implementing a simple _fade out_ animation for the `Alert` component, and I first faced the same issue as @colinmorelli mentioned regarding the need to delay the removal of the component to enable the native CSS animation to perform — considering I was trying to use class toggling to be more compliant with Bootstrap and avoid duplicating information (let's keep a single source of truth).

What I did then is to go full Angular immediately to try avoiding any further issue of this kind (therefore duplicating style definitions in animations definitions) and it worked. But then I wanted to animate an element nested in the template of the component, not the host itself: things were not working anymore. Therefore after multiple tests with different configurations I came to a [plunkr](https://embed.plnkr.co/E39E6tp9AsQgYatK2K07/) demonstrating different cases where __animations on removal__ can work as is (which means without design change) or not.

Please read the content of the `readme.md` file which describes four complementary use cases with only the first one working (even though you will notice that even in this case __we are lacking information__ on the parent side: look at when the `reset` button appears), and also adds an __API/design proposal__ to be able to handle more use cases (not related to animations only) for components "closing".

To sum up quickly: 

- when removing an element from the DOM through a `ngIf` or `ngFor` the subtree will be kept only if this element has an animation _on leave_ attached to it. Therefore, it's not easy to animate only children of a component. But also it's not possible to ensure that the user component (the one including ours) will actually put the structural directive on our component directly. If it's put on a wrapper for any reason, our animation _on leave_ — even if defined on the host — will not be triggered.
- the API/design proposal is just about adding more steps on closing and giving more (related) information to the parent (user) component, so that it can wait for any process to complete before actually removing the child component. It looks like the following protocol: 
	- child ⇒ parent: I want to close
	- parent: checking if can close or not
	- parent ⇒ child: yes you can start your closing process
	- child ⇒ parent: now starting to close
	- child: doing anything needed during closing, especially animations
	- child ⇒ parent: finished closing, you can remove me
	- parent: actually removing the child
	
Please note that the proposal is independent of the animations, therefore any animation system would be able to work (so using class toggling and DOM events for animations would be a possible implementation).

Please tell me what you think about it: 

- did I miss something?
- does the proposal look fair?
- how could it be improved?
- should things be changed in Angular itself instead? (anyways I believe that the user will always need to be in control of what's happening therefore that would also be a change on the public side of Angular — not only an implementation detail)
- anything else?

Thanks in advance,
Cheers!

**Comment by tproenca** [comment]
Is it possible to leverage ng2-animate (https://github.com/yuyang041060120/ng2-animate) for that?

**Comment by intellix** [comment]
Angular 4.2.0 rc.0 introduces a wave of new animation features  https://github.com/angular/angular/blob/master/CHANGELOG.md#420-rc0-2017-05-19

I imagine that query is the interesting one for ng-bootstrap 

The PR with more info: https://github.com/angular/angular/pull/15574

**Comment by emilio-martinez** [comment]
I believe 4.2.0-rc.0 introduces the much-awaited animation query, as well as restoring the programmatic animation that was lost in the renderer since 4.0. Both will probably support a good deal in the animations needed for ng-bootstrap.

**Comment by zulianrizki** [comment]
I think it's great to implement angular 4 animations because it have event like before animate, after animate, staggering etc. That would be a really nice addition to ng bootstrap

**Comment by pavimus** [comment]
Now angular have stable version 4.2.5, is there any progress on this issue?

Now i'm deciding what to use, ng-bootstrap or ngx-bootstrap. At first glance i prefer ng-bootstrap, but ngx-bootstrap have animations (see http://valor-software.com/ngx-bootstrap/index-bs4.html#/modals)

**Comment by Eugeny** [comment]
At least the modal animation should be extremely easy to do by controlling the `show` class on the `ngb-modal-window`, see: https://github.com/ng-bootstrap/ng-bootstrap/issues/894#issuecomment-312850981

**Comment by pkozlowski-opensource** [comment]
Quick update: after considering several different options we are going to do the following:
* try to use Angular animations;
* channel feedback back to the Angular team if we hit any roadblocks / pain points;
* if the roadblocks can't be resolved in Angular itself than we are going to consider other options (CSS animations, JS animations, ...)

**Comment by drmnick** [comment]
Was battling navbar collapse/expand animation and implemented it in similar way to [twbs](https://github.com/twbs/bootstrap/blob/v4-dev/js/src/collapse.js#L193). I think part of this could be migrated to the collapse directive, here is [plnkr](http://plnkr.co/edit/5cLAZ5mz4ZvYhfRCU1hR)

**Comment by MickL** [comment]
Since Angular 4.3 i guess all animations can be done nicely with little effort. 

Please update this thread so everyone can see which animations are still unfinished. I may do one or two for practice.

**Comment by warapitiya** [comment]
Is this still blocked by `Angular Issue` or `WIP`?

> Label says it's blocked. Still valid?

**Comment by pkozlowski-opensource** [comment]
For anyone tracking this issue: thnx to @mattlewis92 awesome job we do have now an experimental branch (https://github.com/ng-bootstrap/ng-bootstrap/tree/animations) where the work on animations is being done. If you are interested in the animations support you might check this branch as well as PRs in the animations milestone:

Review of the existing code / PRs and test-driving this code would be more than appreciated! If you care about animations support in ng-bootstrap, please step in and help out. Thnx!

**Comment by js-digital** [comment]
It seems like you don't support NG-Bootstrap with new features like Animations. 
It's awful. 

Do you consider a contribution to your project? Maybe someone can help you with existing tasks

**Comment by edriang** [comment]
The problem is that in modal window component host class is defined like this:
` '[class]': '"modal fade show" + (windowClass ? " " + windowClass : "")',`

Here, `fade` transition will apply when `show` class is added, but if `show` is already added then none transition is applied, because it is already on final state.

So, easiest solution to support same animation as regular bootstrap4 is to add `show`  class to host after some tick. I think doing that on `afterViewInit` or just using `setTimeout` will do the trick

**Comment by chelaruc** [comment]
Hi, guys! Thank you for the good work! Have you a stable version with animations? Best, Ciprian :D

**Comment by gabestev** [comment]
I agree with @edriang explanation, perhaps you can add some timeout when opening and closing modal. It sequence is matter.

i once did added bootstrap modal manually like this. hope this help for the moment
```
  @ViewChild('myModal) myModal: ElementRef;

  constructor() { }
  
  // your other code 

  openMyModa() {
    // set display style to block
    this._renderer.setStyle(this.myModal.nativeElement, 'display', 'block');

    // after 300 ms add show class
    Observable.interval(300).take(1).subscribe(() => {
      this._renderer.addClass(this.yourModal.nativeElement, 'show');
    });

  closeMyModal() {
    // remove show class
    this._renderer.removeClass(this.myModal.nativeElement, 'show');

    // after 300 ms set display style to none
    Observable.interval(300).take(1).subscribe(() => {
      this._renderer.setStyle(this.yourModal.nativeElement, 'display', 'none');
    });
  }

```

**Comment by kevintanhongann** [comment]
Hi guys. What is the temporary solution to the animations for open and closing the modal? Can I use something like attaching animations on the modal component itself to get the results that I want?

https://www.youtube.com/watch?v=mVqQDEtRBwA

**Comment by markaweb** [comment]
Hello guys, trying to use bootstrap v4 installed via npm and ng-bootstrap v1 with angular v5.

When trying to open modal window, found that there are some missing classes. For example checking modal example in plunker (http://plnkr.co/edit/?p=preview) with added bootstrap.css via link, there is .fade.show class with opacity:1; from _transitions.scss. 

When checking my modal , styles from _transition.scss are missing and modal is opened but not visible.. 
The same issue exists for button styles from _button.scss . 

If I will remove bootstrap package and will add bootstrap v4 via link <rel="stylesheet"> all is working fine  and styles on my site are slightly different (for example fonts and etc).. 

Could you please help to understand, if this is some kind of error or am I missing something? 

Thanks a lot .

**Comment by MickL** [comment]
Ng-bootstrap is not using the css animations provided by bootstrap. Instead the goal is to recode them using the Angular animations module.

It is not implemented yet because there were blocking issues on Angular itself.

**Comment by GuerricPhalippou** [comment]
Is there any estimated time for implementing bootstrap animations in Angular ? I am starting to use this library but the modals and collapse missing effects really lowered my interest :(

**Comment by adrianbarbe** [comment]
That's one of the reason why I have chosen VueJs 2 over Angular. I like Angular but the conservatism and the lack of things like these (animations in UI-packages etc) makes me sad.

**Comment by MickL** [comment]
Actually the Angular animations are pretty awesome. Or can you add an animation in vue before something gets destroyed?

Anyway this project could have used the css animations from Bootstrap but they didnt. Its a decision from ng-bootstrap team and has nothing todo with Angular itself.

**Comment by kapsonic** [comment]
Is there any chance for this being resolved in near future ?

**Comment by kukac7** [comment]
wondering for this feature in the 2.0.0 version?

**Comment by tkhalane** [comment]
Any progress on this very important issue team?

**Comment by pkozlowski-opensource** [comment]
For anyone asking / looking for the "official answer from the maintainers": we've found out that the current Angular animation system has limitations that doesn't allow us to implement animations without resorting to ugly hacks. Instead of implementing hacks we are working with the Angular team (actually, I'm part of the Angular core team) on improving the existing animation system to the point where implementing animations for ng-bootstrap is possible.

Unfortunately things take time as not everything in software development is straightforward - especially when moving in the uncharted territory.  This item remains our top priority but the ultimate resolution will depend on the timing of changes to the Angular animation system.

**Comment by kwall2004** [comment]
Is it not possible to implement animations the same way that Angular Material does?

**Comment by Eugeny** [comment]
I think this is the case of a strive for perfection killing the product. We all know well enough that at least things like modals don't require anything more than class addition/removal to take advantage of the animations already provided by Bootstrap itself, but ng-bootstrap makes it impossible as pointed out by https://github.com/ng-bootstrap/ng-bootstrap/issues/295#issuecomment-339417598

**Comment by hylobatidae** [comment]
Hopefully we can use animations soon. Created a couple of apps using ng-bootstrap and the client always asks about animations to which I consistently reply: "Don't worry, it'll come soon, it's around the corner". 

Reminds me of my experience with ui-bootstrap when it first came out. We all used it, it was abandoned for years and we all moved to AngularStrap. Are there any good alternatives right now?

**Comment by dallinwright** [comment]
For anyone coming to this looking for animation support from this module specifically, it is a WIP for a long time now but Angular itself can now solve a lot of issues for you. I found the solution to what I was looking for (animated navbar dropdown since angular and bootstrap don't always play nice) here:

https://angular.io/guide/animations

Animations are provided in @angular/animations and they can do a lot of stuff now.

**Comment by fbasso** [comment]
I started to have a look at animations in ngBootstrap, and have a try by just trying to add/remove classes as bootstrap do. To me, it is the better way to handle them (if feasible, of course !), as it's completely in the spirit of bootstrap : having solid css bases for the application, then being completely customizable by re-building locally bootstrap or by overriding the classes with a theme. If we go with angular animation, we will lose an easy way to customize the final application and then will come questions as how to change the delay, the timing function, ...

Having say that, I just had a try to go this way with the alert and modal here : https://github.com/fbasso/ng-bootstrap/tree/animation

It's the same idea as @colinmorelli commented it before : adding/removing bootstrap classes when needed and the transitionend event to manage the timing. It seems to work quite well (at the moment, I disabled the animation in the tests, to be sure that nothing is broken with the new code, but they will need to be refactored for sure), the behavior is exactly the same as on bootstrap.

Maybe this solution will not be applicable for all the widgets (I think about the collapse for example, where the panel height needs to be calculated), but for the cases where it is, it's a worthy solution.

**Comment by kultigin** [comment]
I have been studying angular animations, I can animate around the application but I can not put triggers in ng-bootstrap templates. It gives error. I wanted to animate accordion panels. Using :enter and :leave transitions but I can not do it without really going into ng-bootstrap modules and I am not good for that. Do you have any examples of anyone writing animations for accordion? And is it possible to use angular animations without updating to angular 6? We couldn't update our application for two years because of constant rewriting of angular.

**Comment by pkozlowski-opensource** [comment]
Initial design of work in progress: https://gist.github.com/fbasso/4e8e779071c365c976641eb7ec27cfd5

**Comment by gufigueiredo** [comment]
How is the progress on this?

**Comment by tanekim77** [comment]
ng-bootstrap animation would make my day

**Comment by tanekim77** [comment]
Is there a very rough guess on the estimate date of at least a barely working draft?

**Comment by tbickley-mediabowl** [comment]
It looks like these changes don't currently include NgbModalBackdrop? I'd imagine it just needs the same changes as NgbModalWindow though.

**Comment by laserus** [comment]
@pkozlowski-opensource It seems that ng-bootstrap alerts support fade-in / fade-out animations, but the documentation website is silent about it: https://ng-bootstrap.github.io/#/components/alert/examples

Could it be updated? It is impossible to find any info on the usage.

**Comment by aningbo** [comment]
2.5 years later... its like the contributors have given up!

**Comment by KyleCartmell** [comment]
@pkozlowski-opensource @fbasso Can you provide guidance on who owns this concern and the most effective manner in which support could be provided? The "initial design" seems to indicate that we're partially blocked on a refactor but the conversation seems to have gone stale. If we had some idea of what needed doing and what hoops to jump through to ensure that a PR would be accepted without significant rework I could offer some development effort.

**Comment by asvishnyakov** [comment]
Any progress on this?

**Comment by KingDarBoja** [comment]
Hello,

I know this issue has been around for like 3 years and there hasn't been much progress besides [one of maintainers answer](https://github.com/ng-bootstrap/ng-bootstrap/issues/295#issuecomment-382014741) but so far I was able to implement a `ZoomIn` animation using `@HostBinding()` decorator on one of my modal components.

```ts
export const ZoomInOutAnimation: AnimationTriggerMetadata = trigger('ZoomInOut', [
    transition(':enter', [group(
        [
            style({
                transform: 'scale(1.25)',
                opacity: 0,
            }),
            animate('500ms ease-in-out', style({
                transform: 'none',
                opacity: 1,
            }))
        ]
    )]),
    transition(':leave', [group(
        [
            style({
                opacity: 1,
                transform: 'none',
            }),
            animate('500ms ease-out', style({
                opacity: 0,
                transform: 'scale(1.25)',
            }))
        ]
    )])
])
```

However, I noticed that *:leave* transition doesn't work while closing the modal, I guess if it is related to some internal state being managed by the library itself.

This library is pretty good but been searching how to properly trigger an animation on close.

Cheers!

**Comment by asvishnyakov** [comment]
I created inheritor of carousel in my app and use the following animation:

```typescript
import { Observable, BehaviorSubject, Subscription } from 'rxjs';
import { OnInit, OnDestroy, Component, ChangeDetectionStrategy } from '@angular/core';
import { NgbCarousel, NgbSlide, NgbSlideEvent, NgbSlideEventDirection } from '@ng-bootstrap/ng-bootstrap';
import { trigger, state, style, transition, animate } from '@angular/animations';
import { SlideState } from './slide.state';

@Component({
  selector: 'app-carousel',
  animations: [
    trigger('slide', [
      state('previousLeft', style({
        display: 'none',
        transform: 'translateX(-100%)'
      })),
      state('previousRight', style({
        display: 'none',
        transform: 'translateX(100%)'
      })),
      transition('hidden => activeLeft, previousRight => activeLeft', [
        style({
          display: 'block',
          transform: 'translateX(100%)'
        }),
        animate('0.6s')
      ]),
      transition('hidden => activeRight, previousLeft => activeRight', [
        style({
          display: 'block',
          transform: 'translateX(-100%)'
        }),
        animate('0.6s')
      ]),
      transition('activeLeft => previousLeft, activeRight => previousRight,' +
                 'activeRight => previousLeft, activeLeft => previousRight,' +
                 'active => previousLeft, active => previousRight', [
        style({
          display: 'block'
        }),
        animate('0.6s')
      ])
    ])
  ],
  templateUrl: './carousel.component.html',
  styleUrls: ['./carousel.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class CarouselComponent extends NgbCarousel implements OnInit, OnDestroy {

  private slideStates: { [id: string]: BehaviorSubject<SlideState> } = {};
  private slidesSubscription: Subscription;

  getSlideById(slideId: string): NgbSlide {
     return this.slides.find(slide => slide.id === slideId) as NgbSlide;
  }

  getSlideIndexById(slideId: string): number {
    return this.slides.toArray().indexOf(this.getSlideById(slideId));
  }

  getSlideEventDirection(currentActiveSlideId: string, nextActiveSlideId: string): NgbSlideEventDirection {
    const currentIndex = this.getSlideIndexById(currentActiveSlideId);
    const nextIndex = this.getSlideIndexById(nextActiveSlideId);
    const lastIndex = this.slides.length - 1;
    const firstIndex = 0;

    return currentIndex <= nextIndex &&
      !(currentIndex === firstIndex && nextIndex === lastIndex) ||
      currentIndex === lastIndex && nextIndex === firstIndex
      ? NgbSlideEventDirection.LEFT
      : NgbSlideEventDirection.RIGHT;
  }

  getSlideState(id: string): Observable<SlideState> {
    let result = this.slideStates[id];
    if (!result) {
      result = new BehaviorSubject(this.getSlideStateAtSlide(id, this.activeId));
      this.slideStates[id] = result;
    }
    return result;
  }

  getSlideStateAtSlide(id: string, current: string, prev?: string): SlideState {
    let result: SlideState;
    if (!prev) {
      result = current === id ? SlideState.Active : SlideState.Hidden;
    } else {
      const direction = this.getSlideEventDirection(prev, current);
      switch (direction) {
        case NgbSlideEventDirection.LEFT:
          result = current === id ? SlideState.ActiveLeft : prev === id ? SlideState.PreviousLeft : SlideState.Hidden;
          break;
        case NgbSlideEventDirection.RIGHT:
          result = current === id ? SlideState.ActiveRight : prev === id ? SlideState.PreviousRight : SlideState.Hidden;
          break;
        default:
          result = SlideState.Hidden;
          break;
      }
    }
    return result;
  }

  ngOnInit() {
    this.slidesSubscription = this.slide.subscribe((event: NgbSlideEvent) => {
      for (const slideId of Object.keys(this.slideStates)) {
        const slideState = this.slideStates[slideId];
        slideState.next(this.getSlideStateAtSlide(slideId, event.current, event.prev));
      }
    });
  }

  ngOnDestroy() {
    this.slidesSubscription.unsubscribe();
  }
}
```
```typescript
export enum SlideState {
  Active = 'active',
  ActiveLeft = 'activeLeft',
  ActiveRight = 'activeRight',
  PreviousLeft = 'previousLeft',
  PreviousRight = 'previousRight',
  Hidden = 'hidden'
}
```
```html
<div *ngFor="let slide of slides" class="carousel-item" [class.active]="slide.id === activeId" [@slide]="getSlideState(slide.id) | async">
```

**Comment by ZiluckMichael** [comment]
@benouat @Foxandxss @icfantv @maxokorokov @pkozlowski-opensource @wesleycho Can one of you guys add a little bit more transparency on the status of this issue? Is it something that you guys just don't have the time to get to? If that is the case that is fine, but please don't just leave your users in the dark.

**Comment by tbickley-mediabowl** [comment]
@ZiluckMichael if you check the pull request, it looks like it's just waiting to be reviewed https://github.com/ng-bootstrap/ng-bootstrap/pull/2817

**Comment by kapil-kodewit** [comment]
I have stopped using this library because of this issue and now I use [ngx-bootstrap](https://valor-software.com/ngx-bootstrap/#/). I lost hope on this to ever get solved.

**Comment by ZiluckMichael** [comment]
@tbickley-mediabowl Yes, but it has been waiting to be reviewed for over a year now. That is part of the problem.

**Comment by KevinBassaDevelopment** [comment]
If anyone is still searching on how to get proper animations: I recommend switching to ngx-bootstrap, it can be done very quickly & provides animations without any further need for modification.

**Comment by ahmadmu** [comment]
We also had problems with ngx-bootstrap regarding animations. I think the best way is to build your own components right now if you want animations, or use Material.

**Comment by maxokorokov** [comment]
Hey, a short update regarding animations, as things are more clear now.

We've tried a couple of solutions like using Angular animations and using a custom renderer for transitions in the past. Both didn't work out for various reasons.

At the moment focus is to introduce simple internal tools for transition handling with `transitionend` event, customisable in the same way Bootstrap does it.

That is the next big feature we're hoping to release in ng-bootstrap.

The PR at works is #2817, it should move a lot in the coming days.

**Comment by fbasso** [comment]
For information, the active branch for animations is https://github.com/ng-bootstrap/ng-bootstrap/tree/animations

**Comment by siddiq-rehman** [comment]
After this update, can I use ng-bootstrap v8.00 for angular 8?. I see like it supports only angular 10, But I need animations for my angular 8 project. Any suggestion will be appreciated! Thanks.

**Comment by maxokorokov** [comment]
Closing, see #3747
